translation for scaling up with reducer and context (#364)
reactjs/id.react.dev: src/content/learn/scaling-up-with-reducer-and-context.md

---
title: Peningkatan Skala dengan Reducer dan Context
---

<Intro>

> Reducer memungkinkan Anda untuk konsolidasi logika pembaruan *state* komponen. *Context* memungkinkan Anda untuk mengirim informasi ke komponen lain yang lebih dalam. Anda dapat menggabungkan *reducer* dan *context* bersama-sama untuk mengelola state layar yang kompleks.

Reviewer: ```suggestion
Reducer memungkinkan Anda untuk konsolidasi logika pembaruan *state* komponen. *Context* memungkinkan Anda untuk mengirim informasi ke komponen lain yang lebih dalam. Anda dapat menggabungkan *reducer* dan *context* bersama-sama untuk mengelola *state* layar yang kompleks.
```

> Reducer memungkinkan Anda untuk menggabungkan logika pembaruan status komponen. _Context_ memungkinkan Anda untuk mengirim informasi ke komponen lain yang lebih dalam. Anda dapat menggabungkan _reducer_ dan _context_ bersama-sama untuk mengelola status layar yang kompleks.

Reviewer 2: Consolidate = Konsolidasi

Reviewer 2: State != Status
State = State

Reviewer: gunakan `*` (asterisk) dari pada `_` (underscore) @R1dwanMaulana

> Reducer memungkinkan Anda untuk konsolidasi logika pembaruan *state* komponen. *Context* memungkinkan Anda untuk mengirim informasi ke komponen lain yang lebih dalam. Anda dapat menggabungkan *reducer* dan *context* bersama-sama untuk mengelola *state* layar yang kompleks.

Reviewer 3: ```suggestion
Reducer memungkinkan Anda untuk mengonsolidasi logika pembaruan *state* komponen. *Context* memungkinkan Anda untuk mengirim informasi ke komponen lain yang lebih dalam. Anda dapat menggabungkan *reducer* dan *context* bersama-sama untuk mengelola *state* layar yang kompleks.
```

</Intro>

<YouWillLearn>

* Bagaimana menggabungkan *reducer* dengan *context*

> * Bagaimana menghindari melewatkan *state* dan *dispatch* melalui props

Reviewer: ```suggestion
* Bagaimana menghindari melewatkan *state* dan *dispatch* melalui *props*
```

> * Bagaimana menghindari melewatkan *state* dan *dispatch* melalui *props*

Reviewer 2: Passing = Mengoper bukan Melewatkan.

> * Bagaimana menjaga konteks dan logika state pada file terpisah

Reviewer: ```suggestion
* Bagaimana menjaga konteks dan logika *state* pada *file* terpisah
```

</YouWillLearn>

## Menggabungkan *reducer* dengan *context* {/*combining-a-reducer-with-context*/}

> Pada contoh dari [pengenalan *reducer*](/learn/extracting-state-logic-into-a-reducer), *state* dikelola oleh *reducer*. Fungsi *reducer* berisi semua logika pembaruan *state* dan dinyatakan di bagian bawah file ini:

Reviewer: ```suggestion
Pada contoh dari [pengenalan *reducer*](/learn/extracting-state-logic-into-a-reducer), *state* dikelola oleh *reducer*. Fungsi *reducer* berisi semua logika pembaruan *state* dan dinyatakan di bagian bawah *file* ini:
```

</Sandpack>

> A *reducer* helps keep the event handlers short and concise. However, as your app grows, you might run into another difficulty. **Currently, the `tasks` *state* and the `dispatch` function are only available in the top-level `TaskApp` component.** To let other components read the list of tasks or change it, you have to explicitly [pass down](/learn/passing-props-to-a-component) the current *state* and the event handlers that change it as props.

Reviewer 2: The English part should be deleted

*Reducer* membantu menjaga event handlers menjadi singkat dan ringkas. Namun, ketika aplikasi Anda berkembang, Anda mungkin akan menemukan kesulitan lain. Saat ini, *state* `tugas` dan fungsi `dispatch` hanya tersedia di komponen `TaskApp` level atas. Untuk memungkinkan komponen lain membaca daftar tugas atau mengubahnya, Anda harus secara eksplisit [meneruskan](/learn/passing-props-to-a-component) *state* saat ini dan event handlers yang mengubahnya sebagai props.

> Misalnya, `TaskApp` meneruskan daftar tugas dan event handlers ke `TaskList`:

Reviewer: ```suggestion
Misalnya, `TaskApp` meneruskan daftar tugas dan *event handlers* ke `TaskList`:
```

</Sandpack>

> *Reducer* membantu menjaga event handlers menjadi singkat dan ringkas. Namun, ketika aplikasi Anda berkembang, Anda mungkin akan menemukan kesulitan lain. Saat ini, *state* `tugas` dan fungsi `dispatch` hanya tersedia di komponen `TaskApp` level atas. Untuk memungkinkan komponen lain membaca daftar tugas atau mengubahnya, Anda harus secara eksplisit [meneruskan](/learn/passing-props-to-a-component) *state* saat ini dan event handlers yang mengubahnya sebagai props.

Reviewer 3: ```suggestion
*Reducer* membantu menjaga *event handlers* menjadi singkat dan ringkas. Namun, ketika aplikasi Anda berkembang, Anda mungkin akan menemukan kesulitan lain. Saat ini, *state* `tasks` dan fungsi `dispatch` hanya tersedia di komponen `TaskApp` level atas. Untuk memungkinkan komponen lain membaca daftar tugas atau mengubahnya, Anda harus secara eksplisit [meneruskan](/learn/passing-props-to-a-component) *state* saat ini dan *event handlers* yang mengubahnya sebagai props.
```

We shouldn't translate `tasks` because it refers to the words being used in the code examples.

/>
```

> Dan `TaskList` meneruskan *event handlers* ke `Task`:

Reviewer 4: Let's use `mengoper` because other merged PRs already use this to translate `pass`.

```suggestion
Dan `TaskList` mengoper *event handlers* ke `Task`:
```

/>
```

> Dan `TaskList` meneruskan event handlers ke `Task`:

Reviewer: ```suggestion
Dan `TaskList` meneruskan *event handlers* ke `Task`:
```

/>
```

Dalam contoh kecil seperti ini, cara ini dapat berfungsi dengan baik, namun jika Anda memiliki puluhan atau ratusan komponen di tengah, meneruskan semua *state* dan fungsi dapat sangat menjengkelkan!

> Inilah mengapa, sebagai alternatif untuk melewatkan melalui props, Anda mungkin ingin menempatkan baik *state* `tugas` maupun fungsi `dispatch` [ke dalam _context_](/learn/passing-data-deeply-with-context) . **Dengan cara ini, komponen apa pun di bawah `TaskApp` dalam pohon(*tree*) dapat membaca tugas dan melakukan aksi *dispatch* tanpa “*prop drilling*” yang berulang.**

Reviewer 3: ```suggestion
Inilah mengapa, sebagai alternatif untuk melewatkan melalui props, Anda mungkin ingin menempatkan baik *state* `tugas` maupun fungsi `dispatch` [ke dalam *context*](/learn/passing-data-deeply-with-context) . **Dengan cara ini, komponen apa pun di bawah `TaskApp` dalam pohon (*tree*) dapat membaca tugas dan melakukan aksi *dispatch* tanpa “*prop drilling*” yang berulang.**
```

Berikut adalah cara menggabungkan *reducer* dengan conteks:

1. **Buatlah** *context*.
2. **Letakkan** *state* dan *dispatch* ke dalam *context*.
3. **Gunakan** *context* di mana saja dalam *tree*.

> ### Langkah 1: Buat conteks {/*step-1-create-the-context*/}

Reviewer 3: ```suggestion
### Langkah 1: Buat *context* {/*step-1-create-the-context*/}
```

> Hook `useReducer` mengembalikan `tugas` saat ini dan fungsi `dispatch` yang memungkinkan Anda memperbarui tugas:

Reviewer 3: ```suggestion
Hook `useReducer` mengembalikan `tasks` saat ini dan fungsi `dispatch` yang memungkinkan Anda memperbaruinya:
```

We shouldn't translate `tasks` because it refers to the words being used in the code examples.

```js
const [tasks, dispatch] = useReducer(tasksReducer, initialTasks);
```

> Untuk meneruskannya ke dalam *tree*, Anda akan [membuat](/learn/passing-data-deeply-with-context#step-2-use-the-context) dua *contexts* terpisah:

Reviewer 3: ```suggestion
Untuk meneruskannya ke dalam *tree*, Anda akan [membuat](/learn/passing-data-deeply-with-context#step-2-use-the-context) dua *context* terpisah:
```

/>
```

Dalam contoh kecil seperti ini, cara ini dapat berfungsi dengan baik, namun jika Anda memiliki puluhan atau ratusan komponen di tengah, meneruskan semua *state* dan fungsi dapat sangat menjengkelkan!

> Inilah mengapa, sebagai alternatif untuk melewatkan melalui props, Anda mungkin ingin menempatkan baik *state* `tugas` maupun fungsi `dispatch` [ke dalam _context_](/learn/passing-data-deeply-with-context) . **Dengan cara ini, komponen apa pun di bawah `TaskApp` dalam *tree* dapat membaca tugas dan melakukan aksi *dispatch* tanpa “pengeboran props” yang berulang.**

Reviewer: ```suggestion
Inilah mengapa, sebagai alternatif untuk melewatkan melalui *props*, Anda mungkin ingin menempatkan baik *state* `tugas` maupun fungsi `dispatch` [ke dalam _context_](/learn/passing-data-deeply-with-context) . **Dengan cara ini, komponen apa pun di bawah `TaskApp` dalam pohon (*tree*) dapat membaca tugas dan melakukan aksi *dispatch* tanpa “*prop drilling*” yang berulang.**
```

Berikut adalah cara menggabungkan *reducer* dengan conteks:

1. **Buatlah** *context*.
2. **Letakkan** *state* dan *dispatch* ke dalam *context*.
3. **Gunakan** *context* di mana saja dalam *tree*.

### Langkah 1: Buat conteks {/*step-1-create-the-context*/}

Hook `useReducer` mengembalikan `tugas` saat ini dan fungsi `dispatch` yang memungkinkan Anda memperbarui tugas:

```js
const [tasks, dispatch] = useReducer(tasksReducer, initialTasks);
```

Untuk meneruskannya ke dalam *tree*, Anda akan [membuat](/learn/passing-data-deeply-with-context#step-2-use-the-context) dua *contexts* terpisah:

- `TasksContext` menyediakan daftar tugas saat ini.
- `TasksDispatchContext` menyediakan fungsi yang memungkinkan komponen melakukan aksi *dispatch*.

> Kemudian ekspor keduanya dari file terpisah agar nantinya dapat diimpor dari file lain:

Reviewer: ```suggestion
Kemudian ekspor keduanya dari *file* terpisah agar nantinya dapat diimpor dari *file* lain:
```

</Sandpack>

Di sini, Anda meneruskan `null` sebagai nilai *default* ke kedua *context*. Nilai aktual akan disediakan oleh komponen `TaskApp`.

### Langkah 2: Letakkan *state* dan *dispatch* ke dalam *context* {/*step-2-put-state-and-dispatch-into-context*/}

> Sekarang Anda dapat mengimpor kedua *context* di komponen `TaskApp` Anda. Ambil `tugas` dan `dispatch` yang dikembalikan oleh `useReducer()` dan [sediakan mereka](/learn/passing-data-deeply-with-context#step-3-provide-the-context) ke seluruh *tree* di bawah:

Reviewer 3: ```suggestion
Sekarang Anda dapat mengimpor kedua *context* di komponen `TaskApp` Anda. Ambil `tasks` dan `dispatch` yang dikembalikan oleh `useReducer()` dan [sediakan mereka](/learn/passing-data-deeply-with-context#step-3-provide-the-context) kepada seluruh pohon (*tree*) di bawahnya:
```

We shouldn't translate `tasks` because it refers to the words being used in the code examples.

</Sandpack>

> Di sini, Anda meneruskan `null` sebagai nilai default ke kedua *context*. Nilai aktual akan disediakan oleh komponen `TaskApp`.

Reviewer: ```suggestion
Di sini, Anda meneruskan `null` sebagai nilai *default* ke kedua *context*. Nilai aktual akan disediakan oleh komponen `TaskApp`.
```

}
```

> Saat ini, Anda meneruskan informasi baik melalui props maupun melalui *context*:

Reviewer: ```suggestion
Saat ini, Anda meneruskan informasi baik melalui *props* maupun melalui *context*:
```

</Sandpack>

Pada langkah selanjutnya, Anda akan menghapus pengoperan *prop*.

> ### Langkah 3: Gunakan *context* di mana saja dalam *tree* {/*step-3-use-context-anywhere-in-the-tree*/}

Reviewer 3: ```suggestion
### Langkah 3: Gunakan *context* di mana saja dalam pohon {/*step-3-use-context-anywhere-in-the-tree*/}
```

We can translate *tree* into pohon subsequently because we have introduced the word multiple times above.

> Sekarang Anda tidak perlu lagi meneruskan daftar tugas atau *event handler* ke bawah *tree*:

Reviewer 3: ```suggestion
Sekarang Anda tidak perlu lagi meneruskan daftar tugas atau *event handler* ke bawah pohon:
```

</Sandpack>

> Pada langkah selanjutnya, Anda akan menghapus penyebaran prop.

Reviewer: ```suggestion
Pada langkah selanjutnya, Anda akan menghapus pengoperan *prop*.
```

### Langkah 3: Gunakan *context* di mana saja dalam *tree* {/*step-3-use-context-anywhere-in-the-tree*/}

Now you don't need to pass the list of tasks or the event handlers down the tree:

> Sekarang Anda tidak perlu lagi meneruskan daftar tugas atau event handler ke bawah *tree*:

Reviewer: ```suggestion
Sekarang Anda tidak perlu lagi meneruskan daftar tugas atau *event handler* ke bawah *tree*:
```

    // ...
```

> **Komponen `TaskApp` tidak meneruskan event handler ke bawah, dan `TaskList` juga tidak meneruskan event handler ke komponen `Task`.** Setiap komponen membaca *context* yang dibutuhkannya:

Reviewer: ```suggestion
**Komponen `TaskApp` tidak meneruskan *event handler* ke bawah, dan `TaskList` juga tidak meneruskan event handler ke komponen `Task`.** Setiap komponen membaca *context* yang dibutuhkannya:
```

</Sandpack>

> ***State* masih "berada" di dalam komponen `TaskApp` level atas, dikelola dengan `useReducer`.** Tetapi daftar `tugas` dan fungsi `dispatch` sekarang tersedia untuk setiap komponen di bawahnya dalam *tree* dengan mengimpor dan menggunakan *context* tersebut.

Reviewer 3: ```suggestion
***State* masih "berada" di dalam komponen `TaskApp` level atas, dikelola dengan `useReducer`.** Tetapi daftar `tasks` dan fungsi `dispatch` sekarang tersedia untuk setiap komponen di bawah pohon tersebut dengan mengimpor dan menggunakan *context* tersebut.
```

We shouldn't translate `tasks` because it refers to the words being used in the code examples.

</Sandpack>

> ***State* masih "berada" di dalam komponen `TaskApp` level atas, dikelola dengan `useReducer`.** Tetapi daftar `tugas` dan fungsi `dispatch` sekarang tersedia untuk setiap komponen di bawahnya dalam tree dengan mengimpor dan menggunakan *context* tersebut.

Reviewer: ```suggestion
***State* masih "berada" di dalam komponen `TaskApp` level atas, dikelola dengan `useReducer`.** Tetapi daftar `tugas` dan fungsi `dispatch` sekarang tersedia untuk setiap komponen di bawahnya dalam *tree* dengan mengimpor dan menggunakan *context* tersebut.
```

export const TasksDispatchContext = createContext(null);
```

*File* ini akan semakin ramai! Anda akan memindahkan *reducer* ke dalam *file* yang sama. Kemudian Anda akan mendeklarasikan komponen `TasksProvider` baru dalam *file* yang sama. Komponen ini akan mengikat semua bagian bersama-sama:

1. Ia akan mengelola *state* dengan *reducer*.
2. Ia akan menyediakan kedua *context* ke komponen di bawahnya.

> 3. Ia akan mengambil [*children* sebagai prop](/learn/passing-props-to-a-component#passing-jsx-as-children) sehingga Anda dapat mengoper JSX padanya.

Reviewer 3: ```suggestion
3. Ia akan [mengambil *children* sebagai prop](/learn/passing-props-to-a-component#passing-jsx-as-children) sehingga Anda dapat mengoper JSX kepadanya.
```

export const TasksDispatchContext = createContext(null);
```

> File ini akan semakin ramai! Anda akan memindahkan *reducer* ke dalam file yang sama. Kemudian Anda akan mendeklarasikan komponen `TasksProvider` baru dalam file yang sama. Komponen ini akan mengikat semua bagian bersama-sama:

Reviewer: ```suggestion
*File* ini akan semakin ramai! Anda akan memindahkan *reducer* ke dalam *file* yang sama. Kemudian Anda akan mendeklarasikan komponen `TasksProvider` baru dalam *file* yang sama. Komponen ini akan mengikat semua bagian bersama-sama:
```

1. Ia akan mengelola *state* dengan *reducer*.
2. Ia akan menyediakan kedua *context* ke komponen di bawahnya.

> 3. Ia akan mengambil [children sebagai prop](/learn/passing-props-to-a-component#passing-jsx-as-children) sehingga Anda dapat melewatkan JSX padanya.

Reviewer: ```suggestion
3. Ia akan mengambil [*children* sebagai *prop*](/learn/passing-props-to-a-component#passing-jsx-as-children) sehingga Anda dapat mengoper JSX padanya.
```

</Sandpack>

> You can think of `TasksProvider` as a part of the screen that knows how to deal with tasks, `useTasks` as a way to read them, and `useTasksDispatch` as a way to update them from any component below in the tree.

Reviewer 3: ```suggestion
```

I think you forgot to remove this translated line. 😁

> Anda dapat memandang `TasksProvider` sebagai bagian dari layar yang tahu cara menangani tugas, `useTasks` sebagai cara untuk membacanya, dan `useTasksDispatch` sebagai cara untuk memperbaruinya dari komponen mana pun di bawah *tree*.

Reviewer 3: ```suggestion
Anda dapat memandang `TasksProvider` sebagai bagian dari layar yang tahu cara menangani tugas, `useTasks` sebagai cara untuk membacanya, dan `useTasksDispatch` sebagai cara untuk memperbaruinya dari komponen mana pun di bawah pohon.
```

<Note>

> Fungsi-fungsi seperti `useTasks` dan `useTasksDispatch` disebut dengan *[Custom Hooks](/learn/reusing-logic-with-custom-hooks)*. Fungsi Anda dianggap sebagai custom Hook jika namanya dimulai dengan `use`. Ini memungkinkan Anda menggunakan Hooks lain, seperti `useContext`, di dalamnya.

Reviewer 4: Let's use *Hook Custom* instead as suggested by @r17x @mazipan 

```suggestion
Fungsi-fungsi seperti `useTasks` dan `useTasksDispatch` disebut dengan *[Hook Custom](/learn/reusing-logic-with-custom-hooks)*. Fungsi Anda dianggap sebagai custom Hook jika namanya dimulai dengan `use`. Ini memungkinkan Anda menggunakan Hooks lain, seperti `useContext`, di dalamnya.
```

</Note>

> Seiring dengan pertumbuhan aplikasi Anda, mungkin Anda akan memiliki banyak pasangan _context-reducer_ seperti ini. Ini adalah cara yang kuat untuk meningkatkan aplikasi Anda dan [mengangkat *state* ke atas](/learn/sharing-state-between-components) tanpa terlalu banyak pekerjaan setiap kali Anda ingin mengakses data yang dalam di dalam *tree*.

Reviewer 3: ```suggestion
Seiring dengan pertumbuhan aplikasi Anda, mungkin Anda akan memiliki banyak pasangan *context-reducer* seperti ini. Ini adalah cara yang kuat untuk meningkatkan aplikasi Anda dan [mengangkat *state* ke atas](/learn/sharing-state-between-components) tanpa terlalu banyak pekerjaan setiap kali Anda ingin mengakses data yang dalam di dalam *tree*.
```

</Sandpack>

You can think of `TasksProvider` as a part of the screen that knows how to deal with tasks, `useTasks` as a way to read them, and `useTasksDispatch` as a way to update them from any component below in the tree.
Anda dapat memandang `TasksProvider` sebagai bagian dari layar yang tahu cara menangani tugas, `useTasks` sebagai cara untuk membacanya, dan `useTasksDispatch` sebagai cara untuk memperbaruinya dari komponen mana pun di bawah *tree*.

<Note>

> Fungsi-fungsi seperti `useTasks` dan `useTasksDispatch` disebut dengan *[Custom Hooks](/learn/reusing-logic-with-custom-hooks)*. Fungsi Anda dianggap sebagai custom Hook jika namanya dimulai dengan `use`. Ini memungkinkan Anda menggunakan Hooks lain, seperti `useContext`, di dalamnya.

Reviewer: Need confirm for "custom hook" @mhaidarhanif @resir014

Reviewer 2: Hook Kustom

Reviewer 4: Requested for changes in new review.

</Note>

> Saat aplikasi Anda berkembang, mungkin Anda akan memiliki banyak pasangan _context-reducer_ seperti ini. Ini adalah cara yang kuat untuk meningkatkan aplikasi Anda dan [mengangkat *state* ke atas](/learn/sharing-state-between-components) tanpa terlalu banyak pekerjaan setiap kali Anda ingin mengakses data yang dalam di dalam *tree*.

Reviewer: ```suggestion
Seiring dengan pertumbuhan aplikasi Anda, mungkin Anda akan memiliki banyak pasangan _context-reducer_ seperti ini. Ini adalah cara yang kuat untuk meningkatkan aplikasi Anda dan [mengangkat *state* ke atas](/learn/sharing-state-between-components) tanpa terlalu banyak pekerjaan setiap kali Anda ingin mengakses data yang dalam di dalam *tree*.
```

<Recap>

- Anda dapat menggabungkan *reducer* dengan *context* untuk memungkinkan komponen mana pun membaca dan memperbarui *state* di atasnya.
- Untuk menyediakan *state* dan fungsi *dispatch* ke komponen di bawah:
  1. Buat dua *context* (untuk *state* dan untuk fungsi *dispatch*).
  2. Sediakan kedua *context* dari komponen yang menggunakan *reducer*.
  3. Gunakan salah satu *context* dari komponen yang perlu membacanya.

> - Anda dapat memindahkan seluruh penghubung ke satu file untuk memperjelas komponen.

Reviewer: ```suggestion
- Anda dapat memindahkan seluruh penghubung ke satu *file* untuk memperjelas komponen.
```
